### ISSUE TITLE
RFE: add KillMode=mixed-pgrp and KillMode=pgrp which kills process group of main PID rather than only the main PID

### ISSUE THREAD

**Issue by amluto** [systemd/systemd#31093]
### Component

systemd

### Is your feature request related to a problem? Please describe

I have some old-ish software and I'm trying to run it in a systemd service.  It's fairly well-behaved when run in a pty/tty: the main PID sticks around until the software is done, and it responds to Ctrl-C by initiating a clean shutdown.  A second Ctrl-C will tell it to hurry up and shut down faster.  It leaves no descendant processes around when it's done.

But it does not work right in systemd, and, after some debugging, I figured out why: the main PID doesn't respond to Ctrl-C at all.  Instead, it's child, which is in the same process group, handles Ctrl-C.  Terminals handle Ctrl-C by killing the process _group_ (just like a negative PID in kill(2)), whereas systemd is killing just the process.

I think this issue is only relevant to KillMode=process and KillMode=mixed.  (I'm using the latter.)

### Describe the solution you'd like

I think this could be resolved by an option (KillProcessGroup=yes?) that causes the KillSignal to go to the main process group instead of the just the main process.

### Describe alternatives you've considered

This could even be a new default at some risk of regressions.

This could be a KillMode option, like:

```
KillMode=mixed-pgrp
```

As a workaround, I can do:

```
ExecStop=kill -INT -${MAINPID}
```

but I think this is a bit gross.

**Comment by mbiebl** [comment]
Sounds like `KillMode=control-group`, or am I missing something?

**Comment by amluto** [comment]
The program in question runs a whole pile of subprocesses in their own respective process groups, and it wants to manage killing the subprocesses by itself.  So KillMode=control-group will kill everything and this ends up being undesirable.  KillMode=mixed has the correct effect other than failing to initiate the shutdown process at all due to the signal missing its target.

(I would think this program is being silly, except that literally pressing Ctrl-C works perfectly.)

**Comment by poettering** [comment]
So you are implying that the program spawns of various subprocesses and carefully puts some of them in new process groups and others not? i.e. the C-c handler process is *not* put in a new process group, but some other stuff down the tree is? And you thus want the signal to be delivered to the process group the main process is in?

that's a pretty specialist daemon you have there...

So far we didn't bother with the process group concept at all, left that for shell pipelines.

But I see no reason we couldn't add a new KillMode=mixed-pgrp or so, which does as you say. The implementation is trivial after all.

**Comment by poettering** [comment]
or wait, hmm, I am not sure we can really deliver this, because it's not future proof...

we switched killing over to pidfd_send_signal() after all now, and that does not allow sending signals to process groups of processes iirc.

Maybe you can convince @brauner to add a concept where pidfd_send_signal() supports sending to process groups? It has a flags param after all that could be used to indicate that.

But without that I am not convinced we should add this, because it would mean we'd have to regress to using plain kill() if this mode is selected, and we shouldn't add new concepts with that in 2024 anymore...

**Comment by bluca** [comment]
> But without that I am not convinced we should add this, because it would mean we'd have to regress to using plain kill() if this mode is selected, and we shouldn't add new concepts with that in 2024 anymore...

Indeed, also why is it necessary to use process groups? We have cgroups, and fully support delegation and splitting child processes in separate dynamic scopes via `StartAuxiliaryScope()` now, isn't that sufficient?

**Comment by DaanDeMeyer** [comment]
> Indeed, also why is it necessary to use process groups? We have cgroups, and fully support delegation and splitting child processes in separate dynamic scopes via StartAuxiliaryScope() now, isn't that sufficient?

Ctrl+C sends the signal to the foreground process group. So if you have programs that support running interactively and spawn subprocesses it makes sense to run the subprocesses in a separate process group so when a user presses ctrl+c it only goes to the child process and not the parent process. We do this for every subprocess we spawn in mkosi. So these do make sense in some scenarios.

**Comment by amluto** [comment]
> > But without that I am not convinced we should add this, because it would mean we'd have to regress to using plain kill() if this mode is selected, and we shouldn't add new concepts with that in 2024 anymore...
> 
> Indeed, also why is it necessary to use process groups? We have cgroups, and fully support delegation and splitting child processes in separate dynamic scopes via `StartAuxiliaryScope()` now, isn't that sufficient?

This would entail a radical increase in complexity in the program in question.  Right now it has no dependencies on systemd whatsoever -- in fact, it predates systemd.  And it works fine.  I'm sure I could modify the program to work cleanly under systemd if necessary, but probably not in this complex of a manner.

I'll ping @brauner.

**Comment by amluto** [comment]
Posted here, since Oleg is already working on some almost-related enhancements.

https://lore.kernel.org/all/CALCETrUFDkt+K9zG8mczxzAFy9t-6Mx5Cz-Sx+it6a4nt+O0pg@mail.gmail.com/

**Comment by amluto** [comment]
> So you are implying that the program spawns of various subprocesses and carefully puts some of them in new process groups and others not? i.e. the C-c handler process is _not_ put in a new process group, but some other stuff down the tree is? And you thus want the signal to be delivered to the process group the main process is in?

Yes.

> 
> that's a pretty specialist daemon you have there...

I'm guessing my program does roughly what it sounds like mkosi is doing -- it runs its own subprocesses that it wants to manage on its own, so it puts them into their own process groups specifically so that Ctrl-C does _not_ automatically kill them.  Instead it handles Ctrl-C itself so that it can coordinate a clean shutdown of all the subprocesses.

(The reason for the weird internal structure is that the actual main PID is a subreaper, and for simplicity it's set up so that the main PID is _only_ a subreaper and everything else interesting happens in the child.  So the child coordinates the clean shutdown.  This all works very reliably, and it works fine without manipulating cgroups and without any privilege.)

> 
> So far we didn't bother with the process group concept at all, left that for shell pipelines.
> 
> But I see no reason we couldn't add a new KillMode=mixed-pgrp or so, which does as you say. The implementation is trivial after all.

**Comment by Werkov** [comment]
> the main PID doesn't respond to Ctrl-C at all

This seems to be the bug. Main PID should be representative of the service.
If this is some legacy stuff, it could be `Type=forking` + `PIDFile=` so that systemd knows what to signal.

> I see no reason we couldn't add a new KillMode=mixed-pgrp 

Process groups are orthogonal to cgroups, it could in theory mean killing processes of unrelated services (if pgrp members were deliberately or accidentally migrated so). And the we would end up with another unwanted `KillMode=` value.

OT
> splitting child processes in separate dynamic scopes via StartAuxiliaryScope() now

Or `StartTransientUnit()` since old days too. (Both are privileged though so not advisable in this case I guess.)
